corehq/ex-submodules/casexml/apps/case/models.py

```diff
@@ -583,7 +585,13 @@ def _safe_replace_and_force_to_string(me, attr, val):
     
     def update_from_case_update(self, case_update, xformdoc, other_forms=None):
         other_forms = other_forms or {}
-        if case_update.has_referrals():
+        if is_deprecation(xformdoc):
+            # remove the form from actions and xform_ids (based on orig_id)
+            # and rebuild
+            self.xform_ids = [id for id in self.xform_ids if id != xformdoc.orig_id]
+            self.actions = [a for a in self.actions if a.xform_id != xformdoc.orig_id]
```

Comment:
is this approximately what you had in mind? https://github.com/dimagi/commcare-hq/commit/9133144d319be2d7b6249d0f7ef96d741c1fb755

after writing that i decided I wanted to write a test to make sure it's valid so feel free to hold off on merge for a bit while I work on that.